core/src/main/clojure/xtdb/operator/table.clj

```diff
@@ -133,33 +134,22 @@
 
                         (vw/rel-wtr->rdr out-rel)))))}))
 
-(defn- emit-col-table [col-spec table-expr {:keys [param-fields schema] :as opts}]
+(defn- emit-col-table [col-spec table-expr {:keys [param-fields] :as opts}]
   (let [[out-col v] (first col-spec)
         param-types (update-vals param-fields types/field->col-type)
-
         expr (expr/form->expr v (assoc opts :param-types param-types))
         input-types (assoc opts :param-types param-types)
-        projection-spec (expr/->expression-projection-spec out-col expr input-types)
-        field (-> (.getField projection-spec)
-                  (types/unnest-field)
-                  (types/field-with-name (str out-col)))]
-
-    {:fields (-> {(symbol (.getName field)) field}
+        {:keys [field-type !list-expr-fn]} (expr-list/compile-list-expr expr input-types)
+        named-field (types/field-with-name field-type (str out-col))]
+    {:fields (-> {(symbol (.getName named-field)) named-field}
                  (restrict-cols table-expr))
-
      :->out-rel (fn [{:keys [allocator ^RelationReader args]}]
-                  (util/with-open [list-rdr (.project projection-spec allocator (vr/rel-reader [] 1) schema args)]
-                    (let [list-rdr-type (.getType (.getField list-rdr))
-                          list-rdr (cond
-                                     (instance? ArrowType$Null list-rdr-type) nil
-                                     (instance? ArrowType$Union list-rdr-type) (.legReader list-rdr "list")
-                                     :else list-rdr)]
-
-                      (util/with-close-on-catch [el-rdr (.copy (or (some-> list-rdr .getListElements (.withName (str out-col)))
-                                                                   (vr/vec->reader (ZeroVector. (str out-col))))
-                                                               allocator)]
-
-                        (vr/rel-reader [el-rdr] (.getValueCount el-rdr))))))}))
+                  (if (instance? ArrowType$Null (.getType named-field))
+                    (vr/rel-reader [(vr/vec->reader (ZeroVector. (str out-col)))])
+                    (let [^ListExpression list-expr (!list-expr-fn args)
+                          out-vec-writer (vw/->writer (.createVector named-field allocator))]
```

Comment:
need some resource handling here - `createVector` creates a resource that might leak if any of the rest of this function throws an exception.

have a look at `util/with-close-on-catch`? (and lmk if it's unclear what to do :slightly_smiling_face:)